### PR TITLE
Add backfill on websocket reconnections

### PR DESCRIPTION
**Pull request by dphilipson**

### CODE REVIEW

File: src/subscriptions/subscriptionBackfill.ts [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+    toBlockExclusive: number,
+  ): Promise<NewHeadsEvent[]> {
+    const batchParts: BatchPart[] = [];
+    for (let i = fromBlockInclusive; i < toBlockExclusive; i++) {
+      batchParts.push({
+        method: "eth_getBlockByNumber",
+        params: [toHex(i), false],
+      });
+    }
+    const heads = await senders.sendBatch(batchParts);
```
[the remark this answers is not in the document]
Comment (dphilipson) [reply src/subscriptions/subscriptionBackfill.ts]: Yeah, this is something I need to do, and I need to think about the correct behavior. I'm a bit worried about somebody closing their laptop for a week, and then when they come back making thousands of get block requests. Should we declare bankruptcy after a certain number of missed blocks?

File: src/web3-adapter/alchemyContext.ts [hunk 2]
```diff
@@ -24,7 +24,7 @@ export function makeAlchemyContext(
     const provider = makeAlchemyHttpProvider(sendPayload);
     return { provider, setWriteProvider };
   } else if (/^wss?:\/\//.test(url)) {
-    const ws = new SturdyWebSocket(url);
+    const ws = new SturdyWebSocket(url, { debug: true });
```
[the remark this answers is not in the document]
Comment (dphilipson) [reply src/web3-adapter/alchemyContext.ts]: Oh yeah, oops. Thanks!

File: src/subscriptions/subscriptionBackfill.ts [hunk 3]
```diff
@@ -0,0 +1,196 @@
+import { BatchPart, JsonRpcSenders } from "../util/jsonRpc";
+
+export interface NewHeadsEvent {
```
[the remark this answers is not in the document]
Comment (dphilipson) [reply src/subscriptions/subscriptionBackfill.ts]: Oh thanks, this is good to know. I'll do a pass to make sure what we have here matches our server responses. I think I must have gotten these from Infura responses.

File: src/subscriptions/subscriptionBackfill.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (jaypaik) [reply src/subscriptions/subscriptionBackfill.ts]: Cool, 120 blocks sgtm! We can adjust as necessary later

**Pull request merged**